src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -173,7 +173,11 @@ internal JsonConverter<TTarget> CreateCastingConverter<TTarget>()
 
         internal static bool ShouldFlush(Utf8JsonWriter writer, ref WriteStack state)
         {
-            // If surpassed flush threshold then return false which will flush stream.
+            // If surpassed flush threshold then return true which will flush stream.
+            if (state.PipeWriter is { CanGetUnflushedBytes: true } pipeWriter)
```

Comment:
I think I agree with @davidfowl that we should check `CanGetUnflushedBytes` early and throw if not supported. Can implement an opt-in scheme _if_ there is user demand. This can just be a debug assert here.